python/ray/serve/_private/proxy_state.py

```diff
@@ -504,66 +448,78 @@ def update(self, draining: bool = False):
-            except Exception:
-                self.try_update_status(ProxyStatus.UNHEALTHY)
-                logger.warning(
-                    "Unexpected error occurred when checking readiness of "
-                    f"proxy on node {self._node_id}:\n{traceback.format_exc()}"
-                )
-            return
+        else:
+            # At this point, the proxy is either in HEALTHY or DRAINING status.
+            assert self._status in {ProxyStatus.HEALTHY, ProxyStatus.DRAINING}
 
-        # At this point, the proxy is either in HEALTHY or DRAINING status.
-        assert self._status in {ProxyStatus.HEALTHY, ProxyStatus.DRAINING}
+            should_check_health = self._last_health_check_time is None or (
+                self._timer.time() - self._last_health_check_time
+                >= PROXY_HEALTH_CHECK_PERIOD_S
+            )
+            # Perform health-check for proxy's actor (if necessary)
+            if should_check_health:
+                is_healthy_response = self._actor_proxy_wrapper.is_healthy(
+                    PROXY_HEALTH_CHECK_TIMEOUT_S
+                )
+                if is_healthy_response is not None:
+                    if is_healthy_response:
+                        # At this stage status is either HEALTHY or DRAINING, and here
+                        # we simply reset the status
+                        self.try_update_status(self._status)
+                    else:
+                        self.try_update_status(ProxyStatus.UNHEALTHY)
+
+                    self._last_health_check_time = self._timer.time()
+
+            # Handle state transitions (if necessary)
+            if self._status == ProxyStatus.UNHEALTHY:
+                return
+            elif self._status == ProxyStatus.HEALTHY:
+                if draining:
+                    logger.info(
+                        f"Start draining the proxy actor on node {self._node_id}"
+                    )
+                    assert self._last_drain_check_time is None
+
+                    self._actor_proxy_wrapper.update_draining(draining=True)
+                    self.try_update_status(ProxyStatus.DRAINING)
+            elif self._status == ProxyStatus.DRAINING:
+                if not draining:
+                    logger.info(
+                        f"Stop draining the proxy actor on node {self._node_id}"
+                    )
+                    self._last_drain_check_time = None
 
-        self._health_check()
-        if self._status == ProxyStatus.UNHEALTHY:
-            return
+                    self._actor_proxy_wrapper.update_draining(draining=False)
+                    self.try_update_status(ProxyStatus.HEALTHY)
+                else:
+                    should_check_drain = self._last_drain_check_time is None or (
+                        self._timer.time() - self._last_drain_check_time
+                        >= PROXY_DRAIN_CHECK_PERIOD_S
+                    )
+                    if should_check_drain:
```

Comment:
Similar to `should_check_health`, I think we can always just check drain here and relay on the timeout in `is_drained()`.

Non-blocker: Also, just noted this is the only one with a default time. Would be nice to have all 3 checks all have defaults or none has one.